pcs/pcs.8
.TH PCS "8" "April 2021" "pcs 0.9.170" "System Administration Utilities"
.SH NAME
pcs \- pacemaker/corosync configuration system
.SH SYNOPSIS
.B pcs
[\fI\-f file\fR] [\fI\-h\fR] [\fIcommands\fR]...
.SH DESCRIPTION
Control and configure pacemaker and corosync.
.SH OPTIONS
.TP
\fB\-h\fR, \fB\-\-help\fR
Display usage and exit.
.TP
\fB\-f\fR file
Perform actions on file instead of active CIB.
.TP
\fB\-\-debug\fR
Print all network traffic and external commands run.
.TP
\fB\-\-version\fR
Print pcs version information. List pcs capabilities if \fB\-\-full\fR is specified.
.TP
\fB\-\-request\-timeout\fR=<timeout>
Timeout for each outgoing request to another node in seconds. Default is 60s.
.SS "Commands:"
.TP
cluster
 Configure cluster options and nodes.
.TP
resource
 Manage cluster resources.
.TP
stonith
 Manage fence devices.
.TP
constraint
 Manage resource constraints.
.TP
property
 Manage pacemaker properties.
.TP
acl
 Manage pacemaker access control lists.
.TP
qdevice
 Manage quorum device provider on the local host.
.TP
quorum
 Manage cluster quorum settings.
.TP
booth
 Manage booth (cluster ticket manager).
.TP
status
 View cluster status.
.TP
config
 View and manage cluster configuration.
.TP
pcsd
 Manage pcs daemon.
.TP
node
 Manage cluster nodes.
.TP
alert
 Manage pacemaker alerts.
.TP
client
 Manage pcsd client configuration.
.SS "resource"
.TP
[show [<resource id>] | \fB\-\-full\fR | \fB\-\-groups\fR | \fB\-\-hide\-inactive\fR]
Show all currently configured resources or if a resource is specified show the options for the configured resource.  If \fB\-\-full\fR is specified, all configured resource options will be displayed.  If \fB\-\-groups\fR is specified, only show groups (and their resources).  If \fB\-\-hide\-inactive\fR is specified, only show active resources.
.TP
list [filter] [\fB\-\-nodesc\fR]
Show list of all available resource agents (if filter is provided then only resource agents matching the filter will be shown). If \fB\-\-nodesc\fR is used then descriptions of resource agents are not printed.
.TP
describe [<standard>:[<provider>:]]<type> [\fB\-\-full\fR]
Show options for the specified resource. If \fB\-\-full\fR is specified, all options including advanced ones are shown.
.TP
create <resource id> [<standard>:[<provider>:]]<type> [resource options] [\fBop\fR <operation action> <operation options> [<operation action> <operation options>]...] [\fBmeta\fR <meta options>...] [\fBclone\fR [<clone options>] | \fBmaster\fR [<master options>] | \fB\-\-group\fR <group id> [\fB\-\-before\fR <resource id> | \fB\-\-after\fR <resource id>] | \fBbundle\fR <bundle id>] [\fB\-\-disabled\fR] [\fB\-\-no\-default\-ops] [\fB\-\-wait\fR[=n]]
Create specified resource. If \fBclone\fR is used a clone resource is created. If \fBmaster\fR is specified a master/slave resource is created. If \fB\-\-group\fR is specified the resource is added to the group named. You can use \fB\-\-before\fR or \fB\-\-after\fR to specify the position of the added resource relatively to some resource already existing in the group. If \fBbundle\fR is specified, resource will be created inside of the specified bundle. If \fB\-\-disabled\fR is specified the resource is not started automatically. If \fB\-\-no\-default\-ops\fR is specified, only monitor operations are created for the resource and all other operations use default settings. If \fB\-\-wait\fR is specified, pcs will wait up to 'n' seconds for the resource to start and then return 0 if the resource is started, or 1 if the resource has not yet started. If 'n' is not specified it defaults to 60 minutes.

Example: Create a new resource called 'VirtualIP' with IP address 192.168.0.99, netmask of 32, monitored everything 30 seconds, on eth2: pcs resource create VirtualIP ocf:heartbeat:IPaddr2 ip=192.168.0.99 cidr_netmask=32 nic=eth2 op monitor interval=30s
.TP
delete <resource id|group id|master id|clone id>
Deletes the resource, group, master or clone (and all resources within the group/master/clone).
.TP
enable <resource id>... [\fB\-\-wait\fR[=n]]
Allow the cluster to start the resources. Depending on the rest of the configuration (constraints, options, failures, etc), the resources may remain stopped. If \fB\-\-wait\fR is specified, pcs will wait up to 'n' seconds for the resources to start and then return 0 if the resources are started, or 1 if the resources have not yet started. If 'n' is not specified it defaults to 60 minutes.
.TP
disable <resource id>... [\fB\-\-safe\fR [\fB\-\-no\-strict\fR]] [\fB\-\-simulate\fR [\fB\-\-brief\fR]] [\fB\-\-wait\fR[=n]]
Attempt to stop the resources if they are running and forbid the cluster from starting them again. Depending on the rest of the configuration (constraints, options, failures, etc), the resources may remain started.
.br
If \fB\-\-safe\fR is specified, no changes to the cluster configuration will be made if other than specified resources would be affected in any way.
.br
If \fB\-\-no\-strict\fR is specified, no changes to the cluster configuration will be made if other than specified resources would get stopped or demoted. Moving resources between nodes is allowed.
.br
If \fB\-\-simulate\fR is specified, no changes to the cluster configuration will be made and the effect of the changes will be printed instead. If \fB\-\-brief\fR is also specified, only a list of affected resources will be printed.
.br
If \fB\-\-wait\fR is specified, pcs will wait up to 'n' seconds for the resources to stop and then return 0 if the resources are stopped or 1 if the resources have not stopped. If 'n' is not specified it defaults to 60 minutes.
.TP
safe\-disable <resource id>... [\fB\-\-no\-strict\fR] [\fB\-\-simulate\fR [\fB\-\-brief\fR]] [\fB\-\-wait\fR[=n]] [\fB\-\-force\fR]
Attempt to stop the resources if they are running and forbid the cluster from starting them again. Depending on the rest of the configuration (constraints, options, failures, etc), the resources may remain started. No changes to the cluster configuration will be made if other than specified resources would be affected in any way.
.br
If \fB\-\-no\-strict\fR is specified, no changes to the cluster configuration will be made if other than specified resources would get stopped or demoted. Moving resources between nodes is allowed.
.br
If \fB\-\-simulate\fR is specified, no changes to the cluster configuration will be made and the effect of the changes will be printed instead. If \fB\-\-brief\fR is also specified, only a list of affected resources will be printed.
.br
If \fB\-\-wait\fR is specified, pcs will wait up to 'n' seconds for the resources to stop and then return 0 if the resources are stopped or 1 if the resources have not stopped. If 'n' is not specified it defaults to 60 minutes.
.br
If \fB\-\-force\fR is specified, checks for safe disable will be skipped.
.TP
restart <resource id> [node] [\fB\-\-wait\fR=n]
Restart the resource specified. If a node is specified and if the resource is a clone or master/slave it will be restarted only on the node specified.  If \fB\-\-wait\fR is specified, then we will wait up to 'n' seconds for the resource to be restarted and return 0 if the restart was successful or 1 if it was not.
.TP
debug\-start <resource id> [\fB\-\-full\fR]
This command will force the specified resource to start on this node ignoring the cluster recommendations and print the output from starting the resource.  Using \fB\-\-full\fR will give more detailed output.  This is mainly used for debugging resources that fail to start.
.TP
debug\-stop <resource id> [\fB\-\-full\fR]
This command will force the specified resource to stop on this node ignoring the cluster recommendations and print the output from stopping the resource.  Using \fB\-\-full\fR will give more detailed output.  This is mainly used for debugging resources that fail to stop.
.TP
debug\-promote <resource id> [\fB\-\-full\fR]
This command will force the specified resource to be promoted on this node ignoring the cluster recommendations and print the output from promoting the resource.  Using \fB\-\-full\fR will give more detailed output.  This is mainly used for debugging resources that fail to promote.
.TP
debug\-demote <resource id> [\fB\-\-full\fR]
This command will force the specified resource to be demoted on this node ignoring the cluster recommendations and print the output from demoting the resource.  Using \fB\-\-full\fR will give more detailed output.  This is mainly used for debugging resources that fail to demote.
.TP
debug\-monitor <resource id> [\fB\-\-full\fR]
This command will force the specified resource to be monitored on this node ignoring the cluster recommendations and print the output from monitoring the resource.  Using \fB\-\-full\fR will give more detailed output.  This is mainly used for debugging resources that fail to be monitored.
.TP
move <resource id> [destination node] [\fB\-\-master\fR] [lifetime=<lifetime>] [\fB\-\-wait\fR[=n]]
Move the resource off the node it is currently running on by creating a \-INFINITY location constraint to ban the node.  If destination node is specified the resource will be moved to that node by creating an INFINITY location constraint to prefer the destination node.  If \fB\-\-master\fR is used the scope of the command is limited to the master role and you must use the master id (instead of the resource id).  If lifetime is specified then the constraint will expire after that time, otherwise it defaults to infinity and the constraint can be cleared manually with 'pcs resource clear' or 'pcs constraint delete'.  If \fB\-\-wait\fR is specified, pcs will wait up to 'n' seconds for the resource to move and then return 0 on success or 1 on error.  If 'n' is not specified it defaults to 60 minutes.  If you want the resource to preferably avoid running on some nodes but be able to failover to them use 'pcs constraint location avoids'.
.TP
ban <resource id> [node] [\fB\-\-master\fR] [lifetime=<lifetime>] [\fB\-\-wait\fR[=n]]
Prevent the resource id specified from running on the node (or on the current node it is running on if no node is specified) by creating a \-INFINITY location constraint.  If \fB\-\-master\fR is used the scope of the command is limited to the master role and you must use the master id (instead of the resource id).  If lifetime is specified then the constraint will expire after that time, otherwise it defaults to infinity and the constraint can be cleared manually with 'pcs resource clear' or 'pcs constraint delete'.  If \fB\-\-wait\fR is specified, pcs will wait up to 'n' seconds for the resource to move and then return 0 on success or 1 on error.  If 'n' is not specified it defaults to 60 minutes.  If you want the resource to preferably avoid running on some nodes but be able to failover to them use 'pcs constraint location avoids'.
.TP
clear <resource id> [node] [\fB\-\-master\fR] [\fB\-\-expired\fR] [\fB\-\-wait\fR[=n]]
Remove constraints created by move and/or ban on the specified resource (and node if specified). If \fB\-\-master\fR is used the scope of the command is limited to the master role and you must use the master id (instead of the resource id). If \fB\-\-expired\fR is specified, only constraints with expired lifetimes will be removed. If \fB\-\-wait\fR is specified, pcs will wait up to 'n' seconds for the operation to finish (including starting and/or moving resources if appropriate) and then return 0 on success or 1 on error. If 'n' is not specified it defaults to 60 minutes.
.TP
standards
List available resource agent standards supported by this installation (OCF, LSB, etc.).
.TP
providers
List available OCF resource agent providers.
.TP
agents [standard[:provider]]
List available agents optionally filtered by standard and provider.
.TP
update <resource id> [resource options] [op [<operation action> <operation options>]...] [meta <meta operations>...] [\fB\-\-wait\fR[=n]]
Add/Change options to specified resource, clone or multi\-state resource.  If an operation (op) is specified it will update the first found operation with the same action on the specified resource, if no operation with that action exists then a new operation will be created.  (WARNING: all existing options on the updated operation will be reset if not specified.)  If you want to create multiple monitor operations you should use the 'op add' & 'op remove' commands.  If \fB\-\-wait\fR is specified, pcs will wait up to 'n' seconds for the changes to take effect and then return 0 if the changes have been processed or 1 otherwise.  If 'n' is not specified it defaults to 60 minutes.
.TP
op add <resource id> <operation action> [operation properties]
Add operation for specified resource.
.TP
op remove <resource id> <operation action> [<operation properties>...]
Remove specified operation (note: you must specify the exact operation properties to properly remove an existing operation).
.TP
op remove <operation id>
Remove the specified operation id.
.TP
op defaults [options]
Set default values for operations, if no options are passed, lists currently configured defaults. Defaults do not apply to resources which override them with their own defined operations.
.TP
meta <resource id | group id | master id | clone id> <meta options> [\fB\-\-wait\fR[=n]]
Add specified options to the specified resource, group, master/slave or clone.  Meta options should be in the format of name=value, options may be removed by setting an option without a value.  If \fB\-\-wait\fR is specified, pcs will wait up to 'n' seconds for the changes to take effect and then return 0 if the changes have been processed or 1 otherwise.  If 'n' is not specified it defaults to 60 minutes.  Example: pcs resource meta TestResource failure\-timeout=50 stickiness=
.TP
group add <group id> <resource id> [resource id] ... [resource id] [\fB\-\-before\fR <resource id> | \fB\-\-after\fR <resource id>] [\fB\-\-wait\fR[=n]]
Add the specified resource to the group, creating the group if it does not exist. If the resource is present in another group it is moved to the new group. You can use \fB\-\-before\fR or \fB\-\-after\fR to specify the position of the added resources relatively to some resource already existing in the group. By adding resources to a group they are already in and specifying \fB\-\-after\fR or \fB\-\-before\fR you can move the resources in the group. If \fB\-\-wait\fR is specified, pcs will wait up to 'n' seconds for the operation to finish (including moving resources if appropriate) and then return 0 on success or 1 on error. If 'n' is not specified it defaults to 60 minutes.
.TP
group remove <group id> <resource id> [resource id] ... [resource id] [\fB\-\-wait\fR[=n]]
Remove the specified resource(s) from the group, removing the group if no resources remain in it. If \fB\-\-wait\fR is specified, pcs will wait up to 'n' seconds for the operation to finish (including moving resources if appropriate) and then return 0 on success or 1 on error. If 'n' is not specified it defaults to 60 minutes.
.TP
ungroup <group id> [resource id] ... [resource id] [\fB\-\-wait\fR[=n]]
Remove the group (note: this does not remove any resources from the cluster) or if resources are specified, remove the specified resources from the group.  If \fB\-\-wait\fR is specified, pcs will wait up to 'n' seconds for the operation to finish (including moving resources if appropriate) and the return 0 on success or 1 on error.  If 'n' is not specified it defaults to 60 minutes.
.TP
clone <resource id | group id> [clone options]... [\fB\-\-wait\fR[=n]]
Set up the specified resource or group as a clone. If \fB\-\-wait\fR is specified, pcs will wait up to 'n' seconds for the operation to finish (including starting clone instances if appropriate) and then return 0 on success or 1 on error. If 'n' is not specified it defaults to 60 minutes.
.TP
unclone <resource id | group id> [\fB\-\-wait\fR[=n]]
Remove the clone which contains the specified group or resource (the resource or group will not be removed).  If \fB\-\-wait\fR is specified, pcs will wait up to 'n' seconds for the operation to finish (including stopping clone instances if appropriate) and then return 0 on success or 1 on error.  If 'n' is not specified it defaults to 60 minutes.
.TP
master [<master/slave id>] <resource id | group id> [options] [\fB\-\-wait\fR[=n]]
Configure a resource or group as a multi\-state (master/slave) resource.  If \fB\-\-wait\fR is specified, pcs will wait up to 'n' seconds for the operation to finish (including starting and promoting resource instances if appropriate) and then return 0 on success or 1 on error.  If 'n' is not specified it defaults to 60 minutes.  Note: to remove a master you must remove the resource/group it contains.
.TP
bundle create <bundle id> container <container type> [<container options>] [network <network options>] [port\-map <port options>]... [storage\-map <storage options>]... [meta <meta options>] [\fB\-\-disabled\fR] [\fB\-\-wait\fR[=n]]
Create a new bundle encapsulating no resources. The bundle can be used either as it is or a resource may be put into it at any time. If \fB\-\-disabled\fR is specified, the bundle is not started automatically. If \fB\-\-wait\fR is specified, pcs will wait up to 'n' seconds for the bundle to start and then return 0 on success or 1 on error. If 'n' is not specified it defaults to 60 minutes.
.TP
bundle reset <bundle id> [container <container options>] [network <network options>] [port\-map <port options>]... [storage\-map <storage options>]... [meta <meta options>] [\fB\-\-disabled\fR] [\fB\-\-wait\fR[=n]]
Configure specified bundle with given options. Unlike bundle update, this command resets the bundle according given options - no previous options are kept. Resources inside the bundle are kept as they are. If \fB\-\-disabled\fR is specified, the bundle is not started automatically. If \fB\-\-wait\fR is specified, pcs will wait up to 'n' seconds for the bundle to start and then return 0 on success or 1 on error. If 'n' is not specified it defaults to 60 minutes.
.TP
bundle update <bundle id> [container <container options>] [network <network options>] [port\-map (add <port options>) | (remove <id>...)]... [storage\-map (add <storage options>) | (remove <id>...)]... [meta <meta options>] [\fB\-\-wait\fR[=n]]
Add, remove or change options to specified bundle. If you wish to update a resource encapsulated in the bundle, use the 'pcs resource update' command instead and specify the resource id.  If \fB\-\-wait\fR is specified, pcs will wait up to 'n' seconds for the operation to finish (including moving resources if appropriate) and then return 0 on success or 1 on error.  If 'n' is not specified it defaults to 60 minutes.
.TP
manage <resource id>... [\fB\-\-monitor\fR]
Set resources listed to managed mode (default). If \fB\-\-monitor\fR is specified, enable all monitor operations of the resources.
.TP
unmanage <resource id>... [\fB\-\-monitor\fR]
Set resources listed to unmanaged mode. When a resource is in unmanaged mode, the cluster is not allowed to start nor stop the resource. If \fB\-\-monitor\fR is specified, disable all monitor operations of the resources.
.TP
defaults [options]
Set default values for resources, if no options are passed, lists currently configured defaults. Defaults do not apply to resources which override them with their own defined values.
.TP
cleanup [<resource id>] [\fB\-\-node\fR <node>] [\fB\-\-strict\fR]
Make the cluster forget failed operations from history of the resource and re\-detect its current state. This can be useful to purge knowledge of past failures that have since been resolved.
.br
If the named resource is part of a group, or one numbered instance of a clone or bundled resource, the clean\-up applies to the whole collective resource unless \fB\-\-strict\fR is given.
.br
If a resource id is not specified then all resources / stonith devices will be cleaned up.
.br
If a node is not specified then resources / stonith devices on all nodes will be cleaned up.
.TP
refresh [<resource id>] [\fB\-\-node\fR <node>] [\fB\-\-strict\fR]
Make the cluster forget the complete operation history (including failures) of the resource and re\-detect its current state. If you are interested in forgetting failed operations only, use the 'pcs resource cleanup' command.
.br
If the named resource is part of a group, or one numbered instance of a clone or bundled resource, the clean\-up applies to the whole collective resource unless \fB\-\-strict\fR is given.
.br
If a resource id is not specified then all resources / stonith devices will be refreshed.
.br
If a node is not specified then resources / stonith devices on all nodes will be refreshed.
.TP
failcount show [<resource id> [<node> [<operation> [<interval>]]]] [\fB\-\-full\fR]
Show current failcount for resources, optionally filtered by a resource, node, operation and its interval. If \fB\-\-full\fR is specified do not sum failcounts per resource and node. Operation, interval and \fB\-\-full\fR options require pacemaker\-1.1.18 or newer.
.TP
failcount reset [<resource id> [<node> [<operation> [<interval>]]]]
Reset failcount for specified resource on all nodes or only on specified node. This tells the cluster to forget how many times a resource has failed in the past. This may allow the resource to be started or moved to a more preferred location. Operation and interval options require pacemaker-1.1.18 or newer.
.TP
relocate dry\-run [resource1] [resource2] ...
The same as 'relocate run' but has no effect on the cluster.
.TP
relocate run [resource1] [resource2] ...
Relocate specified resources to their preferred nodes.  If no resources are specified, relocate all resources.  This command calculates the preferred node for each resource while ignoring resource stickiness.  Then it creates location constraints which will cause the resources to move to their preferred nodes.  Once the resources have been moved the constraints are deleted automatically.  Note that the preferred node is calculated based on current cluster status, constraints, location of resources and other settings and thus it might change over time.
.TP
relocate show

CHANGELOG.md
# Change Log

## [Unreleased]

### Added
- Elliptic curve TLS certificates are now supported in pcsd ([ghissue#123])
- Support for corosync option `totem.block_unlisted_ips` ([rhbz#1720221])

### Fixed
- Node attribute expressions are now correctly reported as not allowed in
  resource defaults rules ([rhbz#1896458])
- Upgreded to jquery 3.6.0 ([rhbz#1882291, rhbz#1886342])

### Changed
- python3-openssl was replaced with python3-cryptography ([rhbz#1927404])

[ghissue#123]: https://github.com/ClusterLabs/pcs/issues/123
[rhbz#1720221]: https://bugzilla.redhat.com/show_bug.cgi?id=1720221
[rhbz#1882291]: https://bugzilla.redhat.com/show_bug.cgi?id=1882291
[rhbz#1886342]: https://bugzilla.redhat.com/show_bug.cgi?id=1886342
[rhbz#1896458]: https://bugzilla.redhat.com/show_bug.cgi?id=1896458
[rhbz#1927404]: https://bugzilla.redhat.com/show_bug.cgi?id=1927404

## [0.10.8] - 2021-02-01

### Added
- Support for changing corosync configuration in an existing cluster
  ([rhbz#1457314], [rhbz#1667061], [rhbz#1856397], [rhbz#1774143])
- Command to show structured corosync configuration (see `pcs cluster
  config show` command) ([rhbz#1667066])

### Fixed
- Improved error message with a hint in `pcs cluster cib-push` ([ghissue#241])
- Option --wait was not working with pacemaker 2.0.5+ ([ghissue#260])
- Explicitly close libcurl connections to prevent stalled TCP connections in
  CLOSE-WAIT state ([ghissue#261], [rhbz#1885841])
- Fixed parsing negative float numbers on command line ([rhbz#1869399])
- Removed unwanted logging to system log (/var/log/messages) ([rhbz#1917286])
- Fixed rare race condition in `pcs cluster start --wait` ([rhbz#1794062])
- Better error message when unable to connect to pcsd ([rhbz#1619818])

### Deprecated
- Commands `pcs config import-cman` and `pcs config export
  pcs-commands|pcs-commands-verbose` have been deprecated ([rhbz#1851335])
- Entering values starting with '-' (negative numbers) without '--' on command
  line is now deprecated ([rhbz#1869399])

[ghissue#241]: https://github.com/ClusterLabs/pcs/issues/241
[ghissue#260]: https://github.com/ClusterLabs/pcs/issues/260
[ghissue#261]: https://github.com/ClusterLabs/pcs/issues/261
[rhbz#1457314]: https://bugzilla.redhat.com/show_bug.cgi?id=1457314
[rhbz#1619818]: https://bugzilla.redhat.com/show_bug.cgi?id=1619818
[rhbz#1667061]: https://bugzilla.redhat.com/show_bug.cgi?id=1667061
[rhbz#1667066]: https://bugzilla.redhat.com/show_bug.cgi?id=1667066
[rhbz#1774143]: https://bugzilla.redhat.com/show_bug.cgi?id=1774143
[rhbz#1794062]: https://bugzilla.redhat.com/show_bug.cgi?id=1794062
[rhbz#1851335]: https://bugzilla.redhat.com/show_bug.cgi?id=1851335
[rhbz#1856397]: https://bugzilla.redhat.com/show_bug.cgi?id=1856397
[rhbz#1869399]: https://bugzilla.redhat.com/show_bug.cgi?id=1869399
[rhbz#1885841]: https://bugzilla.redhat.com/show_bug.cgi?id=1885841
[rhbz#1917286]: https://bugzilla.redhat.com/show_bug.cgi?id=1917286


## [0.10.7] - 2020-09-30

### Added
- Support for multiple sets of resource and operation defaults, including
  support for rules ([rhbz#1222691], [rhbz#1817547], [rhbz#1862966],
  [rhbz#1867516], [rhbz#1869399])
- Support for "demote" value of resource operation's "on-fail" option
  ([rhbz#1843079])
- Support for 'number' type in rules ([rhbz#1869399])
- It is possible to set custom (promotable) clone id in `pcs resource create`
  and `pcs resource clone/promotable` commands ([rhbz#1741056])

### Fixed
- Prevent removing non-empty tag by removing tagged resource group or clone
  ([rhbz#1857295])
- Clarify documentation for 'resource move' and 'resource ban' commands with
  regards to the 'lifetime' option.
- Allow moving both promoted and demoted promotable clone resources
  ([rhbz#1875301])

### Deprecated
- `pcs resource [op] defaults <name>=<value>...` commands are deprecated now.
  Use `pcs resource [op] defaults update <name>=<value>...` if you only manage
  one set of defaults, or `pcs resource [op] defaults set` if you manage
  several sets of defaults. ([rhbz#1817547])

[rhbz#1222691]: https://bugzilla.redhat.com/show_bug.cgi?id=1222691
[rhbz#1741056]: https://bugzilla.redhat.com/show_bug.cgi?id=1741056
[rhbz#1817547]: https://bugzilla.redhat.com/show_bug.cgi?id=1817547
[rhbz#1843079]: https://bugzilla.redhat.com/show_bug.cgi?id=1843079
[rhbz#1857295]: https://bugzilla.redhat.com/show_bug.cgi?id=1857295
[rhbz#1862966]: https://bugzilla.redhat.com/show_bug.cgi?id=1862966
[rhbz#1867516]: https://bugzilla.redhat.com/show_bug.cgi?id=1867516
[rhbz#1869399]: https://bugzilla.redhat.com/show_bug.cgi?id=1869399
[rhbz#1875301]: https://bugzilla.redhat.com/show_bug.cgi?id=1875301


## [0.10.6] - 2020-06-11

### Security
- Web UI sends HTTP headers: Content-Security-Policy, X-Frame-Options and
  X-Xss-Protection

### Added
- When creating a cluster, verify the cluster name does not prevent mounting
  GFS2 volumes ([rhbz#1782553])
- An option to run 'pcs cluster setup' in a local mode (do not connect to any
  nodes, save corosync.conf to a specified file) ([rhbz#1839637])
- Support for pacemaker tags. Pcs provides commands for creating and removing
  tags, adding and/or removing IDs to/from tags, and listing current tag
  configuration.  ([rhbz#1684676])
- Support for tag ids in commands resource enable/disable/manage/unmanage
  ([rhbz#1684676])
- `pcs resource [safe-]disable --simulate` has a new option `--brief` to print
  only a list of affected resources ([rhbz#1833114])

### Fixed
- Keep autogenerated IDs of set constraints reasonably short ([rhbz#1387358],
  [rhbz#1824206])
- Pcs is now compatible with Ruby 2.7 and Python 3.8. To achieve this, it newly
  depends on python3-distro package.
- `pcs status` works on remote nodes again (broken since pcs-0.10.4)
  ([rhbz#1830552])
- Fixed inability to create colocation constraint from web ui ([rhbz#1832973])
- Actions going through pcsd no longer time out after 30s (broken since
  pcs-0.10.5) ([rhbz#1833506])


pcs/pcs.py
#!/usr/bin/env python

from __future__ import absolute_import
from __future__ import division
from __future__ import print_function
from __future__ import unicode_literals

import sys
import os
import getopt

import usage
import cluster
import resource
import stonith
import prop
import constraint
import acl
import utils
import status
import settings
import config
import pcsd
import node


usefile = False
filename = ""
def main(argv):
    utils.subprocess_setup()
    global filename, usefile
    orig_argv = argv[:]
    utils.pcs_options = {}
    modified_argv = []
    real_argv = []
    try:
        # we change --cloneopt to "clone" for backwards compatibility
        new_argv = []
        for arg in argv:
            if arg == "--cloneopt" or arg == "--clone":
                new_argv.append("clone")
            elif arg.startswith("--cloneopt="):
                new_argv.append("clone")
                new_argv.append(arg.split('=',1)[1])
            else:
                new_argv.append(arg)
        argv = new_argv

        # we want to support optional arguments for --wait, so if an argument
        # is specified with --wait (ie. --wait=30) then we use them
        waitsecs = None
        new_argv = []
        for arg in argv:
            if arg.startswith("--wait="):
                tempsecs = arg.replace("--wait=","")
                if len(tempsecs) > 0:
                    waitsecs = tempsecs
                    arg = "--wait"
            new_argv.append(arg)
        argv = new_argv

        # h = help, f = file,
        # p = password (cluster auth), u = user (cluster auth),
        # V = verbose (cluster verify)
        pcs_short_options = "hf:p:u:V"
        pcs_long_options = [
            "debug", "version", "help", "fullhelp",
            "force", "autocorrect", "interactive", "autodelete",
            "all", "full", "groups", "local", "wait", "config",
            "start", "enable", "disabled", "off",
            "pacemaker", "corosync",
            "no-default-ops", "defaults", "nodesc",
            "clone", "master", "name=", "group=", "node=",
            "from=", "to=", "after=", "before=",
            "transport=", "rrpmode=", "ipv6",
            "addr0=", "bcast0=", "mcast0=", "mcastport0=", "ttl0=", "broadcast0",
            "addr1=", "bcast1=", "mcast1=", "mcastport1=", "ttl1=", "broadcast1",
            "wait_for_all=", "auto_tie_breaker=", "last_man_standing=",
            "last_man_standing_window=",
            "token=", "token_coefficient=", "consensus=", "join=",
            "miss_count_const=", "fail_recv_const=",
            "corosync_conf=", "cluster_conf=",
            "remote",
            #in pcs status - do not display resorce status on inactive node
            "hide-inactive",
        ]
        # pull out negative number arguments and add them back after getopt
        prev_arg = ""
        for arg in argv:
            if len(arg) > 0 and arg[0] == "-":
                if arg[1:].isdigit() or arg[1:].startswith("INFINITY"):
                    real_argv.append(arg)
                else:
                    modified_argv.append(arg)
            else:
                # If previous argument required an argument, then this arg
                # should not be added back in
                if not prev_arg or (not (prev_arg[0] == "-" and prev_arg[1:] in pcs_short_options) and not (prev_arg[0:2] == "--" and (prev_arg[2:] + "=") in pcs_long_options)):
                    real_argv.append(arg)
                modified_argv.append(arg)
            prev_arg = arg

        pcs_options, argv = getopt.gnu_getopt(modified_argv, pcs_short_options, pcs_long_options)
    except getopt.GetoptError as err:
        print(err)
        usage.main()
        sys.exit(1)
    argv = real_argv
    for o, a in pcs_options:
        if not o in utils.pcs_options:
            utils.pcs_options[o] = a
        else:
            # If any options are a list then they've been entered twice which isn't valid
            utils.err("%s can only be used once" % o)
        if o == "-h" or o == "--help":
            if len(argv) == 0:
                usage.main()
                sys.exit()
            else:
                argv = [argv[0], "help" ] + argv[1:]
        elif o == "-f":
            usefile = True
            filename = a
            utils.usefile = usefile
            utils.filename = filename
        elif o == "--corosync_conf":
            settings.corosync_conf_file = a
        elif o == "--cluster_conf":
            settings.cluster_conf_file = a
        elif o == "--version":
            print(settings.pcs_version)
            sys.exit()
        elif o == "--fullhelp":
            usage.full_usage()
            sys.exit()
        elif o == "--wait":
            utils.pcs_options[o] = waitsecs

    if len(argv) == 0:
        usage.main()
        sys.exit(1)

    command = argv.pop(0)
    if (command == "-h" or command == "help"):
        usage.main()
        return
    cmd_map = {
        "resource": resource.resource_cmd,
        "cluster": cluster.cluster_cmd,
        "stonith": stonith.stonith_cmd,
        "property": prop.property_cmd,
        "constraint": constraint.constraint_cmd,
        "acl": acl.acl_cmd,
        "status": status.status_cmd,
        "config": config.config_cmd,
        "pcsd": pcsd.pcsd_cmd,
        "node": node.node_cmd,
    }
    if command not in cmd_map:
        usage.main()
        sys.exit(1)
    # root can run everything directly, also help can be displayed,
    # working on a local file also do not need to run under root
    if (os.getuid() == 0) or (argv and argv[0] == "help") or usefile:
        cmd_map[command](argv)
        return
    # specific commands need to be run under root account, pass them to pcsd
    # don't forget to allow each command in pcsd.rb in "post /run_pcs do"
    root_command_list = [
        ['cluster', 'auth', '...'],
        ['cluster', 'corosync', '...'],
        ['cluster', 'destroy', '...'],
        ['cluster', 'disable', '...'],
        ['cluster', 'enable', '...'],
        ['cluster', 'node', '...'],
        ['cluster', 'pcsd-status', '...'],
        ['cluster', 'setup', '...'],
        ['cluster', 'start', '...'],
        ['cluster', 'stop', '...'],
        ['cluster', 'sync', '...'],
        # ['config', 'restore', '...'], # handled in config.config_restore
        ['pcsd', 'sync-certificates'],
        ['status', 'nodes', 'corosync-id'],
        ['status', 'nodes', 'pacemaker-id'],
        ['status', 'pcsd', '...'],
    ]
    argv_cmd = argv[:]
    argv_cmd.insert(0, command)
    for root_cmd in root_command_list:
        if (
            (argv_cmd == root_cmd)
            or
            (
                root_cmd[-1] == "..."
                and
                argv_cmd[:len(root_cmd)-1] == root_cmd[:-1]
            )
        ):
            # handle interactivity of 'pcs cluster auth'
            if argv_cmd[0:2] == ["cluster", "auth"]:
                if "-u" not in utils.pcs_options:
                    username = utils.get_terminal_input('Username: ')
                    orig_argv.extend(["-u", username])
                if "-p" not in utils.pcs_options:
                    password = utils.get_terminal_password()
                    orig_argv.extend(["-p", password])

            # call the local pcsd
            err_msgs, exitcode, std_out, std_err = utils.call_local_pcsd(
                orig_argv, True
            )
            if err_msgs:
                for msg in err_msgs:
                    utils.err(msg, False)
                sys.exit(1)
            if std_out.strip():
                print(std_out)
            if std_err.strip():
                sys.stderr.write(std_err)
            sys.exit(exitcode)
            return
    cmd_map[command](argv)

if __name__ == "__main__":
    main(sys.argv[1:])
